source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+
+        if( ( xResult = AzureIoTMQTT_Publish( &( xAzureIoTProvisioningClientHandle->_internal.xMQTTContext ),
+                                              &xMQTTPublishInfo, usPublishPacketIdentifier ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "Failed to publish. Error=%d", xResult ) );
+            xReturn = eAzureIoTProvisioningPublishFailed;
+        }
+        else
+        {
+            xReturn = eAzureIoTProvisioningSuccess;
+        }
+
+        prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, xReturn );
+    }
+    else
+    {
+        AZLogWarn( ( "Request action called in wrong state %d\r\n",
+                     xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Implementation of parsing response action, this action is only allowed in azureiotprovisioningWF_STATE_RESPONSE
+ *
+ * */
+static void prvAzureIoTProvisioningClientParseResponse( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle )
+{
+    az_result xCoreResult;
+    az_span xPayload = az_span_create( xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponsePayload,
+                                       ( int32_t ) xAzureIoTProvisioningClientHandle->_internal.xProvisioningLastResponsePayloadLength );
+    az_span xTopic = az_span_create( xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponseTopic,
+                                     ( int32_t ) xAzureIoTProvisioningClientHandle->_internal.usProvisioningLastResponseTopicLength );
+
+    /* Check the state.  */
+    if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_RESPONSE )
+    {
+        xCoreResult =
+            az_iot_provisioning_client_parse_received_topic_and_payload( &xAzureIoTProvisioningClientHandle->_internal.xProvisioningClientCore,
+                                                                         xTopic, xPayload, &xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse );
+
+        if( az_result_failed( xCoreResult ) )
+        {
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningFailed );
+            AZLogError( ( "IoTProvisioning client failed to parse packet, error status: %08x", xCoreResult ) );
+            return;
+        }
+
+        if( xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.operation_status == AZ_IOT_PROVISIONING_STATUS_ASSIGNED )
+        {
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningSuccess );
+        }
+        else if( xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.retry_after_seconds == 0 )
+        {
+            AZLogError( ( "IoTProvisioning client failed with error %d, extended error status: %u\r\n",
```

Comment:
Nit - i'd add a fact that server has 0 for retry so it's clear why we're giving up.